subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.fingerprint.impl;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.file.RelativePath;
+import org.gradle.api.internal.file.AbstractFileTreeElement;
+import org.gradle.api.specs.Spec;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.nativeintegration.filesystem.FileSystem;
+import org.gradle.internal.nativeintegration.filesystem.Stat;
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.SnapshottingFilter;
+import org.gradle.util.GFileUtils;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.function.BiPredicate;
+
+public class PatternSetSnapshottingFilter implements SnapshottingFilter {
+    private final PatternSet patternSet;
+    private final FileSystem fileSystem;
+
+    public PatternSetSnapshottingFilter(PatternSet patternSet, FileSystem fileSystem) {
+        this.fileSystem = fileSystem;
+        this.patternSet = patternSet;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return patternSet.isEmpty();
+    }
+
+    @Override
+    public BiPredicate<FileSystemLocationSnapshot, Iterable<String>> getAsSnapshotPredicate() {
```

Comment:
We could pass in `Stat` here instead of storing it as a field. Not sure if it's better, but maybe?

Reply:
I like that we don't need `Stat` for an arbitrary `SnapshottingFilter`, but only when creating one for `PatternSpec`s. The `DefaultFileSystemSnapshotter` also needs something different to `Stat`, given that we don't have the class available.